### PR TITLE
[DRAFT] Automated testing for Weaviate applications

### PR DESCRIPTION
**Pull request by dandv**
https://649b9093aed035153b9a1cc9--tangerine-buttercream-20c32f.netlify.app/blog/automated-testing

### CODE REVIEW

File: _includes/code/automated-testing.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    'class': class_name,
+    'description': 'A Jeopardy! question',
+    'vectorizer': 'text2vec-openai',
+}
+
+# Clean slate for local testing (GitHub Actions VMs start fresh) because Weaviate data is persisted in embedded mode.
+if client.schema.exists(class_name):
+    client.schema.delete_class(class_name)
+if not client.schema.exists(class_name):
+    client.schema.create_class(class_definition)
```
Comment (sebawita) [inline _includes/code/automated-testing.py:32]: This feels more complicated that it needs to be.

Maybe we can just replace it with:
```
# Clean slate for local testing (GitHub Actions VMs start fresh) because Weaviate data is persisted in embedded mode.
client.schema.delete_all()
```

Then later you could create a new collection, and even do a little assert to verify the collection is there.

Yes, I know that this would be testing Weaviate functionality, but it would be good to start with a small win.
To show how in a few lines of code we can create a new collection and then `assert` that the collection is there.
(in a way, this could be testing if the collection configuration is valid, as one day we might introduce breaking changes and then this test would flag it)

File: blog/2023-06-29-automated-testing/index.mdx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+
+## Sample integration test
+
+To demonstrate the principles above, we'll write an integration test that consists of the following steps:
+
+1. Import a data set into Weaviate using documented best practices ([batching](/developers/weaviate/manage-data/import) and [streaming](/developers/weaviate/manage-data/import#tip-stream-data-from-large-files))
+  a. Have vectorization enabled, so we effectively test a workflow that combines Weaviate and an external API (to control API call costs, we can limit the number of objects imported)
+2. Ensure that all specified data objects have been imported, and no extraneous ones exist
+3. Export back objects along with their vectors (useful for backups and recreating the database without hitting the vectorization provider)
```
Comment (sebawita) [inline blog/2023-06-29-automated-testing/index.mdx:47]: Weaviate has a built-in backup functionality. I don't think we should teach here how to perform a manual backup.

Also, this is not related to automated testing, which is the topic of this blog post.

File: blog/2023-06-29-automated-testing/index.mdx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+* Integration testing should focus on automatically catching common issues including:
+  * Connection or authentication issues with the inference provider (and with Weaviate, if not using it embedded)
+  * Incomplete or incorrect data imports
+  * Using the correct number of dimensions when [bringing your own vectors](/developers/weaviate/tutorials/custom-vectors)
+  * Schema issues, like invalid class names, properties, or data types
+  * Race conditions or concurrency errors from multiple requests
+
+
+## Sample integration test
+
```
Comment (sebawita) [inline blog/2023-06-29-automated-testing/index.mdx:41]: This section seems a bit thin. It basically can be summarised as:
- we will do 1-2-3
- a big chunk of code

Can we break this down into smaller `###segments`, as you walk the reader through your code?
Maybe something like (not exactly these titles):
- Step 1 – connect to Weaviate Embedded + clean up (delete existing schemas)
- Step 2 – create the `JeopardyQuestion` collection and (add this part) assert it exists
- Step 3 – import data and assert the count

Plus we could add Step 4
- Step 4 – a query test – to verify an expected object is returned

**Review by sebawita — CHANGES_REQUESTED** [review]

**Comment by dandv** [comment]
Closing in favor of new branch without conflicts.
